controller: add workqueues/ratelimiting

sjenning opened the pull request:
This PR adds rate limited workqueues to the controller.

https://github.com/kubernetes/client-go/tree/master/util/workqueue

The workqueues allow the reconciliation to happen asynchronously wrt the informer event handlers.  This allows us to reschedule (reinsert into the workqueue) the operation if the reconciliation fails due to transient condition like a resource not existing (yet).  Since we can retry these failed operations at a faster interval than the informer sync interval, the state can converge more quickly.

The rate limiting gives us back-off on the retry interval.

@pmorie @jpeeler @derekwaynecarr

MHBauer commented on cmd/controller-manager/app/options/options.go:
```diff
@@ -55,4 +56,5 @@ func (s *ControllerManagerServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.K8sKubeconfigPath, "k8s-kubeconfig", "./kubeconfig", "Path to k8s core kubeconfig")
 	fs.StringVar(&s.ServiceCatalogAPIServerURL, "service-catalog-api-server-url", "", "The URL for the service-catalog API server")
 	fs.StringVar(&s.ServiceCatalogKubeconfigPath, "service-catalog-kubeconfig", "./servicecatalogkubeconfig", "Path to service-catalog kubeconfig")
+	fs.Int32Var(&s.ConcurrentSyncs, "concurrent-syncs", s.ConcurrentSyncs, "The port that the controller-manager's http service runs on")
```
The comment here is wrong.

derekwaynecarr replied:
do we need to expose concurrent syncs yet, or can we just choose a decent default?

sjenning replied:
i really don't see a reason to expose it.  it is just how it is done in the kube controller-manager.  i can remove it for now.  no reason to expose a knob before we need to.

MHBauer commented on cmd/controller-manager/app/controller_manager.go:
```diff
@@ -300,7 +300,7 @@ func StartControllers(s *options.ControllerManagerServer,
 
 		glog.V(5).Info("Running controller")
-		go serviceCatalogController.Run(stop)
+		go serviceCatalogController.Run(int(s.ConcurrentSyncs), stop)
```
Should this be a configuration to `NewController` instead?

sjenning replied:
elsewhere in the kube code, the number of workers is passed in this way

https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/deployment_controller.go#L143

MHBauer replied:
Any idea why? I don't see why this argument is special. It's still based on a flag like everything else.

MHBauer replied:
I get it now the "controller-manager" vs the "controller". Weird split.

All comments on Run need to be updated to mention the new argument.

arschles commented:
@sjenning what is the rate limiting for here?

MHBauer commented on pkg/controller/controller.go:
```diff
@@ -110,26 +110,45 @@ type controller struct {
+		go wait.Until(c.brokerWorker, time.Second, stopCh)
+		go wait.Until(c.serviceClassWorker, time.Second, stopCh)
+		go wait.Until(c.instanceWorker, time.Second, stopCh)
+		go wait.Until(c.bindingWorker, time.Second, stopCh)
```
How does this wait combine with each worker? Each worker has an infinite loop inside, and wait is going to create one each second. What is the code flow that cleans up workers?

MHBauer replied:
I guess I don't understand how the queues work.

sjenning replied:
the `ShutDown()` calls in the defer block before this do the queue cleanup.  `wait.Until()` is just a way to say, if this goroutine exits for whatever reason (only reason in this case is a program error), restart it.  `wait.Until()` calls the function synchronously so if the function doesn't return, it will not run it again.

MHBauer replied:
The synchronous call is non-obvious. Can you add a comment to that effect?

Is there any purpose to the timeout?

sjenning replied:
It isn't a timeout; it is a retry interval if the function happens to return; which it shouldn't unless it.  Also, if `stopCh` is closed, wait.Until() returns and thus the whole goroutine exits.

The comment would really belong in the godoc for `wait.Until()`.

MHBauer replied:
Right. Not a timeout, very much the opposite. 

Is there a logic behind the choice of one second?

sjenning replied:
Not really :smile: It basically just serves to rate limit the goroutine restart rate if the program gets in a state where the goroutine crashes quickly ever time for some reason.

MHBauer commented:
Is there a document on the queue details and usage?

derekwaynecarr commented:
since i asked @sjenning to do this work, can you expand the PR description for those less familiar with controllers in core kube?

sjenning commented:
@arschles the rate limiting retry lets us retry reconciliation operations at a faster rate than the informer sync for operations the failed the first time through like, for example, if a dependent resource hasn't been created yet.

sjenning commented:
@MHBauer i can't find documentation on the workqueue but the code is here

https://github.com/kubernetes/client-go/tree/master/util/workqueue

Interface is here

https://github.com/kubernetes/client-go/blob/master/util/workqueue/rate_limitting_queue.go#L20-L33

sjenning commented:
@derekwaynecarr I expanded the description.  I guess it was a little terse before :wink:

sjenning commented:
travis failure due to apiserver timeout flake btw

MHBauer commented:
As an aside, is there an equivalent for client-go logic? Or should we somehow client-go in this? It's irritating for users of client-go currently to 'do something, wait a little bit if it fails, and backoff'. We've got a bunch of test code that just wraps `wait.Poll` and friends. This seems to be a more explicit thing that could maybe be integrated directly into the client somehow.

arschles commented:
@sjenning got it, thanks for the explanation.

It sounds like this change begins to address #628 (or fixes?), and also addresses part of #629. Can you state that in the OP?

derekwaynecarr commented:
@arschles -- correct.

@sjenning -- thanks for the expanded details ;-)

pmorie commented:
I'm hesitant to merge this so late in the day as we close out 0.0.2 - rebase and we'll merge on monday.

pmorie commented:
@sjenning if you rebase this we can get it in ASAP and let it bake before the 0.0.3 release

sjenning commented:
@pmorie in progress

sjenning commented:
@pmorie rebased

MHBauer commented on pkg/apiserver/admission/initializer.go:
```diff
@@ -0,0 +1,92 @@
+	kubeInformers kubeinformers.SharedInformerFactory
+}
+
+var _ admission.PluginInitializer = pluginInitializer{}
```
This is a check to make sure our initializer is the same as the upstream one.

sjenning replied:
yikes. not sure how this file got pulled into my commit.  shouldn't be there.

/me goes to look at .gitignore

MHBauer commented on pkg/controller/controller.go:
```diff
@@ -155,8 +194,23 @@ const (
+
 // reconcileBroker is the control-loop that reconciles a Broker.
-func (c *controller) reconcileBroker(broker *v1alpha1.Broker) {
+func (c *controller) reconcileBroker(broker *v1alpha1.Broker) error {
```
Nobody calls this directly anymore. Instead we have one guy put things into a queue, and another guy takes them out to process them. It seems like this processing could be done among multiple controllers as long as it was only done by one controller.

sjenning replied:
True, but there is no easy way to coordinate them without races.  That is why this is not done and only one controller runs at a time.

sjenning replied:
You would need to set up a message queue accessible to all controllers (network accessible) to which events were queued by a watching agent and dequeued by controllers.  I'm having flashbacks of setting up AMQP for Openstack...

MHBauer replied:
Yup. This is a pure comment. No action needed.

sjenning commented on pkg/version/version.go:
```diff
@@ -0,0 +1,43 @@
+		Compiler:  runtime.Compiler,
+		Platform:  fmt.Sprintf("%s/%s", runtime.GOOS, runtime.GOARCH),
+	}
+}
```
same. this file should not be in the commit. my bad!

MHBauer commented on pkg/controller/controller.go:
```diff
@@ -145,6 +164,26 @@ func (c *controller) brokerDelete(obj interface{}) {
+// It enforces that the reconcileBrokerKey is never invoked concurrently with the same key.
+func (c *controller) brokerWorker() {
+	for {
+		func() {
```
Does this func need to be here? It wraps something that is immediately called. Is it so the defer gets run? Should we just put the Done at the end? I don't see any other way out of the loop.

sjenning replied:
Yes, it is just a convenient way to make sure that `Done()` is always called on the dequeue key.

The deployment controller makes this anonymous function into a named one called `processNextWorkItem()` but they are doing the same thing
https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/deployment_controller.go#L460-L476

sjenning replied:
I did it the way the service controller does it: https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/deployment/deployment_controller.go#L460-L476

MHBauer replied:
We *know* where the exit is, and there is only *one* exit, so just call the `Done` there vs in a `defer`.

sjenning replied:
It is true this could be simplified and/or made more clear.

MHBauer commented on pkg/controller/controller.go:
```diff
@@ -110,26 +110,45 @@ type controller struct {
+		c.serviceClassQueue.ShutDown()
+		c.instanceQueue.ShutDown()
+		c.bindingQueue.ShutDown()
+	}()
```
This implicitly depends on `<-stopCh` below, which is the overall signal to shut everything down. Is there a way to explicitly depend on it? Or maybe a comment somehow? I think I just need to read it again a couple more times. There's a couple of moving pieces, but nothing looks wrong to me.

arschles replied:
Instead of putting this in a deferred func, why not put it under the log line that reads:

```go
glog.Info("Shutting down service-catalog controller")
```

Doing so makes these lines explicitly reliant on the `stopCh`

sjenning commented:
@MHBauer ok so looking over this code again, it has some issues.  I fixed them up in seperate commit so you can more easily review the delta.

Basically I fixed:
- The worker function should return when a `quit` is dequeued
- The worker function is actually a little _more_ convoluted now, as it returns a function to be used in `wait.Until()` but now there is just 1 worker function rather than 4
- I was adding the key back in the case of error twice; once in the worker and once in the reconcile*Key functions

pmorie commented:
@sjenning think this will be ready tomorrow, perhaps?

MHBauer commented on pkg/controller/controller.go:
```diff
@@ -41,8 +44,16 @@ import (
+	// a resource is going to be requeued:
+	//
+	// 5ms, 10ms, 20ms, 40ms, 80ms, 160ms, 320ms, 640ms, 1.3s, 2.6s, 5.1s, 10.2s, 20.4s, 41s, 82s
+	maxRetries = 15
```
Really cool! I'd like to have something built into the client or usable for tests as well.

MHBauer replied:
wait.ExponentialBackoff

MHBauer commented on pkg/controller/controller.go:
```diff
@@ -99,7 +108,7 @@ func NewController(
 // Open Service Broker compliant Brokers.
 type Controller interface {
 	// Run runs the controller until the given stop channel can be read from.
```
update comment so mention workers int.

MHBauer commented on pkg/controller/controller.go:
```diff
@@ -110,26 +119,45 @@ type controller struct {
+	bindingQueue           workqueue.RateLimitingInterface
 }
 
 // Run runs the controller until the given stop channel can be read from.
```
Update comment to mention workers int.

MHBauer commented on pkg/controller/controller.go:
```diff
@@ -379,16 +457,31 @@ func (c *controller) updateBrokerFinalizers(
+		return nil
+	}
+	if err != nil {
+		glog.Infof("Unable to retrieve ServiceClass %v from store: %v", key, err)
```
Is this error more urgent to know about? Should it be warning level?

arschles replied:
It should be `glog.Errorf`

arschles commented on pkg/apis/componentconfig/types.go:
```diff
@@ -56,4 +56,9 @@ type ControllerManagerConfiguration struct {
+	// ConcurrentSyncs is the number of service catalog resources that are
+	// allowed to sync concurrently. Larger number = more responsive SC
+	// operations, but more CPU (and network) load.
+	ConcurrentSyncs int32
```
why does this need to be an `int32`?

arschles commented on pkg/controller/controller.go:
```diff
@@ -769,7 +900,7 @@ func (c *controller) reconcileBinding(binding *v1alpha1.Binding) {
 			"The Binding references an Instance which references ServicePlan that does not exist. "+s,
 		)
-		return
+		return fmt.Errorf("service plan not found")
```
can you put the name of the service plan in this error message?

pmorie commented on pkg/controller/controller.go:
```diff
@@ -110,26 +119,45 @@ type controller struct {
 func (c *controller) brokerAdd(obj interface{}) {
-	broker, ok := obj.(*v1alpha1.Broker)
-	if broker == nil || !ok {
+	key, err := cache.DeletionHandlingMetaNamespaceKeyFunc(obj)
```
Nit: I think it would be extremely helpful to newcomers to indicate why we use this particular key func in the first place we use it (which is here) within this file.

pmorie commented on pkg/controller/controller.go:
```diff
@@ -145,6 +173,39 @@ func (c *controller) brokerDelete(obj interface{}) {
 	glog.V(4).Infof("Received delete event for Broker %v", broker.Name)
 }
 
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
```
Nit: I think this should come immediately after `Run` within the file

pmorie commented on pkg/controller/controller.go:
```diff
@@ -700,19 +813,37 @@ func (c *controller) instanceDelete(obj interface{}) {
 
-	c.reconcileBinding(binding)
+func (c *controller) reconcileBindingKey(key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
```
Likewise, I think we should explain the use of `SplitMetaNamespaceKey` the first place it's used, since it's not at all obvious what `SplitMetaNamespace` means.

arschles replied:
We already have logic like this in `./pkg/storage/tpr/keyer.go`

pmorie replied:
> We already have logic like this in ./pkg/storage/tpr/keyer.go

Not sure what your point is here, @arschles.  Elaborate?

sjenning commented:
@pmorie @arschles ok, all comments addressed (to my knowledge).  rebased and ready once again.

pmorie commented:
@sjenning unfortunately, rebase needed again.  I'm going to bump this to 0.0.4 - sorry for the churn.

sjenning commented:
@pmorie rebased

MHBauer commented on pkg/controller/controller.go:
```diff
@@ -111,26 +119,83 @@ type controller struct {
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the reconciler is never invoked concurrently with the same key.
+func worker(queue workqueue.RateLimitingInterface, resourceType string, reconciler func(key string) error) func() {
```
How important is it that we print the resource type? This argument seems to exist solely to be logged. Is there a way to get this info without passing it in?

pmorie replied:
I would prefer to log the resource type.  AFAIK there is not a good way to get that info except to pass it in.

pmorie commented:
@sjenning unfortunately, another rebase needed :(

sjenning commented:
@pmorie it seems that the rebase and merge on this need to be a near atomic operation if this is ever going to merge.  after you get done get done with surgery of course, we can get together and coordinate this.

pmorie commented:
Agree - we can hold off other controller merges when i'm back to get it in.
Sound good?

On Mon, Apr 17, 2017 at 1:58 PM Seth Jennings <notifications@github.com>
wrote:

> @pmorie <https://github.com/pmorie> it seems that the rebase and merge on
> this need to be a near atomic operation if this is ever going to merge.
> after you get done get done with surgery of course, we can get together and
> coordinate this.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes-incubator/service-catalog/pull/648#issuecomment-294543179>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAWXmBBLRGDQp4OSFELxQc3S6HvirNapks5rw6gsgaJpZM4M0qIi>
> .
>

pmorie commented:
@sjenning Let's get this in today - I'm willing to punt all controller merges until after this goes in.

MHBauer commented on pkg/apis/componentconfig/types.go:
```diff
@@ -56,4 +56,9 @@ type ControllerManagerConfiguration struct {
 	// all informers.
 	ResyncInterval time.Duration
+
+	// ConcurrentSyncs is the number of service catalog resources that are
```
This is the number of workers for each resource. As far as I understand it, the "number of service catalog resources " is `concurrent syncs * #of resource types`. Here, by default, 5 * 4 = 20.

pmorie commented on pkg/controller/controller.go:
```diff
@@ -118,29 +126,86 @@ type controller struct {
-	if broker == nil || !ok {
+	// DeletionHandlingMetaNamespaceKeyFunc returns a unique key for the resource obj and
+	// handles the special case where the obj is a DeletedFinalStateUnknown obj which
+	// acts a place holder for object that have been delete from the watch event confirming
```
objects

pmorie replied:
deleted

pmorie replied:
In general this sentence reads a little weird.

sjenning replied:
i'm opened to suggestions that are less that 10 lines of comments long :smile:

sjenning replied:
nevermind, after rereading it, i was missing some words

pmorie commented on pkg/controller/controller.go:
```diff
@@ -830,19 +949,37 @@ func (c *controller) instanceDelete(obj interface{}) {
+	}
+	binding, err := c.bindingLister.Bindings(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		glog.Infof("Binding has been deleted %v", key)
```
Suggest changing this message to `not doing work for binding %v because it has been deleted`

pmorie commented on pkg/controller/controller.go:
```diff
@@ -902,7 +1039,7 @@ func (c *controller) reconcileBinding(binding *v1alpha1.Binding) {
 		)
 		c.recorder.Event(binding, api.EventTypeWarning, errorNonexistentServicePlanReason, s)
-		return
+		return fmt.Errorf("service plan %s not found", instance.Spec.PlanName)
```
Why not re-use the message in `s` here?

pmorie commented on test/integration/controller_test.go:
```diff
@@ -285,6 +285,8 @@ func newTestController(t *testing.T) (
 		t.Fatal(err)
 	}
 
+	go testController.Run(1, nil)
```
comment the nil argument? ie, `nil /* this is what this is */`

sjenning replied:
i'll just move it down a line and use the `stopCh` defined on the next line

arschles commented on pkg/controller/controller.go:
```diff
@@ -118,29 +128,86 @@ type controller struct {
+				defer queue.Done(key)
+
+				err := reconciler(key.(string))
+				if err == nil {
```
2 things:

- Can you ensure that `key` can be asserted to a `string`
- Collapse this to `if err := reconciler(...); err == nil {`

sjenning replied:
1) unless there is a programming error, yes
2) we use err below so the scope needs to extend outside the if block

sjenning replied:
> We already have logic like this in ./pkg/storage/tpr/keyer.go

The cache package is in client-go.  Maybe TPR should just use `SplitMetaNamespaceKey` as well.

Something is weird with github right now and I can't respond to the comment thread where your comment appears :disappointed:

pmorie replied:
I don't understand how the controller uses keys for the work queue and what TPR does are related.

pmorie commented on pkg/controller/controller.go:
```diff
@@ -476,16 +561,31 @@ func (c *controller) updateBrokerFinalizers(
+func (c *controller) reconcileServiceClassKey(key string) error {
+	serviceClass, err := c.serviceClassLister.Get(key)
+	if errors.IsNotFound(err) {
+		glog.Infof("ServiceClass has been deleted %v", key)
```
Make this log message more explanatory.  It's tough to know what this does as someone reading the log who hasn't read the controller code.

arschles commented:
I will merge this after CI passes

arschles commented:
@kibbles-n-bytes it looks like we're seeing the same flake here: 

```console
/var/lib/jenkins/workspace/service-catalog-PR-testing2/src/github.com/kubernetes-incubator/service-catalog/contrib/hack/test_walkthrough.sh: line 147: User provided service broker pod did not come up successfully. (exit 1)
```

arschles commented:
@sjenning can you push another empty/amended commit to trigger another jenkins run? I'd like to see that pass - flakes notwithstanding - before we merge this

kibbles-n-bytes commented:
@arschles I'll run it again

arschles commented:
Thanks @kibbles-n-bytes for working out the flake here.

arschles commented:
Fixes #628
